File: problems/792_number_of_matching_subsequences/solution.py
from typing import List


class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        waiting = collections.defaultdict(list)
        for word in words:
            waiting[word[0]].append(iter(word[1:]))
        for c in s:
            old = waiting[c]
            waiting[c] = []
            for it in old:
                nxt = next(it, None)
                if nxt is None:
                    waiting['#'].append(it)
                else:
                    waiting[nxt].append(it)
        return len(waiting['#'])

File: problems/792_number_of_matching_subsequences/Solution.java
import java.util.*;

class Solution {
    public int numMatchingSubseq(String s, String[] words) {
        List<Deque<Character>> waiting = new ArrayList<>();
        for (int i = 0; i < 128; i++) waiting.add(new ArrayDeque<>());
        List<List<Deque<Character>>> buckets = new ArrayList<>();
        for (int i = 0; i < 128; i++) buckets.add(new ArrayList<>());
        for (String word : words) {
            Deque<Character> deque = new ArrayDeque<>();
            for (char c : word.toCharArray()) deque.add(c);
            buckets.get(deque.peek()).add(deque);
        }
        int count = 0;
        for (char c : s.toCharArray()) {
            List<Deque<Character>> old = buckets.get(c);
            buckets.set(c, new ArrayList<>());
            for (Deque<Character> deque : old) {
                deque.poll();
                if (deque.isEmpty()) count++;
                else buckets.get(deque.peek()).add(deque);
            }
        }
        return count;
    }
}
